graph.cpp

```diff
@@ -10,6 +10,14 @@ double AbstractFactor::eval(const values &/*x*/) { assert("abstract eval" && fal
 values AbstractFactor::gradient_at(const values &/*x*/) { assert("abstract grad" && false); }
 hessian AbstractFactor::hessian_at(const values &/*x*/) { assert("abstract hess" && false); }
 
+// Used to handle graph trimming so graph size does not grow arbitrarily.
+// Shifts all factor indices related to robot poses down by `poseSize`. If this
+// moves the index outside the range used for poses, return false.
+// If this returns false, the factor should be removed from the graph.
+bool AbstractFactor::shiftIndices(int /*poseSize*/, int /*firstPoseIdx*/) {
+  assert("abstract shift" && false);
+}
+
```

Comment:
Similar to my other comment (no idea if this will appear before or after it), if these are supposed to be abstract functions that shouldn't be called, you should just declare them as pure virtual and omit these implementations entirely, making any calls to the abstract functions a compiler error.
```suggestion
```

Reply:
Thanks for pointing this out!! I struggle a bit with virtual methods, especially when template classes are involved.